### PR TITLE
Lw/6664 follow api

### PR DESCRIPTION
**Pull request by johnnieskywalker**
# [Follow API](https://app.shortcut.com/oazo-apps/story/6664/create-api-endpoints-for-followed-vaults-following)
# [Follow button handling](https://app.shortcut.com/oazo-apps/story/7143/follow-vault-using-ui-api)
  
## Changes 👷‍♀️
- GET endpoint to fetch followed vaults
- POST endpoint to Follow vault
- Follow/Following/Unfollow button in VaultHeadline
- Managing FollowButton state depending if vault is being followed, processing or not followed, unfollow when vault is followed and mouse is hovering on button
  
## How to test 🧪
- FollowVaults flag must be turned on
image

- Go to vault page of vault you want to follow, NOTE vault must exist in db otherwise endpoint will return 500, as not being able to create follow record for non-existing vault
image

- Note that when opening vault page, button should be in Loading state when it's checking if vault is followed or not
- Open a vault that is not yet followed by your account (there's no record with this vault in `users_who_follow_vaults` table)
image

- Click follow, check if button is in loading state when waiting for response from server
image

- After following check if record with vault of your preference is added to db and button is in `Following` state
image

- Check if button is switching state to `Unfollow` upon hovering on it with mouse
image

### CODE REVIEW

**Comment by johnnieskywalker** [comment]
mutliple users can follow same vault, its fine 
image

**Comment by johnnieskywalker** [comment]
read endpoint 
image

**Comment by johnnieskywalker** [comment]
post api endpoint for saving works 
image

**Comment by johnnieskywalker** [comment]
image

File: features/termsOfService/jwt.ts [hunk 1]
```diff
@@ -13,7 +13,7 @@ const LOCAL_STORAGE_GNOSIS_SAFE_PENDING = 'gnosis-safe-pending'
 const basePath = getConfig()?.publicRuntimeConfig.basePath || ''
 
 export type JWToken = string
-
+// TODO this function is not used only in tos ~Ł
```
Comment (paszkowskiDamian) [inline features/termsOfService/jwt.ts:16]: I guess no one will ever move it, so maye just do it to `/features/jwt`

File: handlers/follow/follow.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    where: { user_address: { equals: address.toLocaleLowerCase() } },
+  })
+
+  return results
+}
+
+const usersWhoFollowVaultsSchema = z.object({
+  user_address: z.string(),
+  vault_id: z.number(),
+  tos_doc_version: z.string(),
```
Comment (paszkowskiDamian) [inline handlers/follow/follow.ts:20]: tos version?

File: pages/api/follow/[address].tsx [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+
+const handler: NextApiHandler = async (req: NextApiRequest, res: NextApiResponse<any>) => {
+  switch (req.method) {
+    case 'GET':
+      return await getFollow(req, res)
+    default:
+      return res.status(405).end()
+  }
+}
+export default handler
```
Comment (paszkowskiDamian) [inline pages/api/follow/[address].tsx:12]: is it your intention for anyone to check anyone address or it should be just accessible by the owner?

File: features/shared/followApi.ts [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  token: string,
+): Promise<UsersWhoFollowVaults[]> {
+  return fetch(`${basePath}/api/follow`, {
+    method: 'POST',
+    headers: {
+      'Content-Type': 'application/json',
+      authorization: 'Bearer ' + token,
+    },
+    body: JSON.stringify({
+      user_address: followerAddress,
```
Comment (paszkowskiDamian) [inline features/shared/followApi.ts:21]: I think this address should be taken from token, in this case I can set follow/unfollow for anybody.

File: pages/api/follow/[address].tsx (hunk 3 above)
Comment (johnnieskywalker) [reply pages/api/follow/[address].tsx:12]: anyone can check the address, no requirement of restricting it

File: handlers/follow/follow.ts (hunk 2 above)
Comment (johnnieskywalker) [reply handlers/follow/follow.ts:20]: let me know if you think I overcomplicated the model, I can refactor it 

https://github.com/OasisDEX/oasis-borrow/pull/1686/files - this one was fine locally on empty db, but was failing on staging as it was not unique constraint
https://github.com/OasisDEX/oasis-borrow/pull/1709/files - so I added tos_version

**Comment by johnnieskywalker** [comment]
Changed method to `getUserFromRequest(req)`

tested with mainnet vault

image

**Review by paszkowskiDamian — APPROVED** [review]

**Pull request merged**